source/common/router/scoped_rds.cc

```diff
@@ -0,0 +1,269 @@
+#include "common/router/scoped_rds.h"
+
+#include <memory>
+
+#include "envoy/admin/v2alpha/config_dump.pb.h"
+#include "envoy/api/v2/srds.pb.validate.h"
+
+#include "common/common/assert.h"
+#include "common/config/subscription_factory.h"
+
+// Types are deeply nested under Envoy::Config::ConfigProvider; use 'using-directives' across all
+// ConfigProvider related types for consistency.
+using Envoy::Config::ConfigProvider;
+using Envoy::Config::ConfigProviderInstanceType;
+using Envoy::Config::ConfigProviderManager;
+using Envoy::Config::ConfigProviderPtr;
+
+namespace Envoy {
+namespace Router {
+
+namespace ScopedRoutesConfigProviderUtil {
+
+ConfigProviderPtr
+create(const envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager&
+           config,
+       Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
+       ConfigProviderManager& scoped_routes_config_provider_manager) {
+  ASSERT(config.route_specifier_case() == envoy::config::filter::network::http_connection_manager::
+                                              v2::HttpConnectionManager::kScopedRoutes);
+
+  switch (config.scoped_routes().config_specifier_case()) {
+  case envoy::config::filter::network::http_connection_manager::v2::ScopedRoutes::
+      kScopedRouteConfigurationsList: {
+    const envoy::config::filter::network::http_connection_manager::v2::
+        ScopedRouteConfigurationsList& scoped_route_list =
+            config.scoped_routes().scoped_route_configurations_list();
+    ProtobufTypes::ConstMessagePtrVector config_protos(
+        scoped_route_list.scoped_route_configurations().size());
+    for (const auto& it : scoped_route_list.scoped_route_configurations()) {
+      Protobuf::Message* clone = it.New();
+      clone->CopyFrom(it);
```

Comment:
Nit: `MergeFrom`

Reply:
I removed this code (the entirety of the `create()` function) from this PR since it is not yet used and was affecting code coverage stats. I will introduce it again in the next PR where it is actually used.